R/grid.R

```diff
@@ -0,0 +1,295 @@
+#'    grid-template-columns: 100px 2fr 1fr;
+#'    grid-template-areas: 'a a a' 'b b b';
+#'    {custom_style_grid_container}"
+#' }
+#'
+grid_container_css <- function(css_grid_template_areas, rows_height, cols_width) {
+  grid_container_styles <- c(
+    "display: grid",
+    "height: 100%",
+    glue::glue("grid-template-rows: {paste(rows_height, collapse = ' ')}"),
+    glue::glue("grid-template-columns: {paste(cols_width, collapse = ' ')}"),
+    glue::glue("grid-template-areas: {css_grid_template_areas}"),
+    "{custom_style_grid_container}" # To be rendered by glue::glue() later
+  )
+  paste(grid_container_styles, collapse = "; ")
+}
+
+#' Generate list of HTML div elements representing grid areas.
+#'
+#' @param area_names
+#'
+#' @return list of shiny::tags$div
+#'
+#' @details This is a helper function used in grid_template()
+#' \preformatted{
+#'   list_of_area_tags(c("header", "main", "footer"))
+#' }
+#' returns the following list:
+#' \preformatted{
+#'   [[1]] <div style="grid-area: header; {custom_style_grid_area_header}">{{ header }}</div>
+#'   [[2]] <div style="grid-area: main; {custom_style_grid_area_main}">{{ main }}</div>
+#'   [[3]] <div style="grid-area: footer; {custom_style_grid_area_footer}">{{ footer }}</div>
+#' }
+#'
+list_of_area_tags <- function(area_names) {
+  lapply(area_names,
+    function(name) shiny::tags$div(
+     style = as.character(glue::glue("grid-area: {name}; {{custom_style_grid_area_{name}}}")), paste("{{", name, "}}")
+    )
+  )
+}
+
+#' Define a template of a CSS grid
+#'
+#' @param default
+#' Template for desktop:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#' @param mobile
+#' Template for mobile:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#'
+#' @return
+#' list(template = [string], area_names = [vector of strings])
+#'
+#' template - contains template that can be formatted with glue::glue() function
+#'
+#' area_names - contain all unique area names used in grid definition
+#'
+#' @details
+#' Example usage:
+#' \preformatted{
+#'   myGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("header", "header", "header"),
+#'       c("menu",   "main",   "right1"),
+#'       c("menu",   "main",   "right2")
+#'     ),
+#'     rows_height = c("50px", "auto", "100px"),
+#'     cols_width = c("100px", "2fr", "1fr")
+#'   ))
+#'
+#'   display_grid(myGrid)
+#' }
+#' Second example:
+#' \preformatted{
+#'   subGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("top_left", "top_right"),
+#'       c("bottom_left", "bottom_right")
+#'     ),
+#'     rows_height = c("50\%", "50\%"),
+#'     cols_width = c("50\%", "50\%")
+#'   ))
+#'
+#'   display_grid(subGrid)
+#' }
+#'
+#' @export
+#'
+grid_template <- function(
+  default = list(areas = "", rows_height = c(), cols_width = c()),
```

Comment:
are these default parameters any useful here, or they're just placeholders for people to fill in? if the latter, I think they should be removed and just put in the description. `default` is mandatory to specify anyway, isn't it?

Reply:
I refactored it to more concrete argument, so it is indeed a default, not just a placeholder to fill in. Now when you call `grid_template()` you will get a grid with single area 100% x 100% size:

```
default = list(areas = rbind(c("main_area")), rows_height = c("100%"), cols_width = c("100%"))
```